lua/luasnip/util/environ.lua

```diff
@@ -1,190 +1,193 @@
+			name
+		)
+	)
+
+	opts.eager = opts.eager or {}
+	local multiline_vars = opts.multiline_vars or false
+
+	local type_of_it = type(multiline_vars)
+
+	assert(
+		type_of_it == "table"
+			or type_of_it == "boolean"
+			or type_of_it == "function",
+		("Your namespace %s can't have `is_table` of type %s"):format(
+			name,
+			type_of_it
+		)
+	)
+
+	-- If type is function we don't have to override it
+	if type_of_it == "table" then
+		local is_table_set = {}
+
+		for _, key in ipairs(multiline_vars) do
+			is_table_set[key] = true
+		end
 
-function lazy_vars.WORKSPACE_FOLDER() -- The path of the opened workspace or folder
-	return buf_to_ws_part()[1]
-end
+		opts.is_table = function(key)
+			return is_table_set[key] or false
+		end
+	elseif type_of_it == "boolean" then
+		opts.is_table = function(_)
+			return multiline_vars
+		end
+	else -- is a function
+		opts.is_table = multiline_vars
+	end
 
-function lazy_vars.WORKSPACE_NAME() -- The name of the opened workspace or folder
-	local parts = vim.split(buf_to_ws_part()[1] or "", "[\\/]")
-	return parts[#parts]
-end
+	if opts.vars and type(opts.vars) == "table" then
+		opts.vars = tbl_to_lazy_env(opts.vars)
+	end
 
--- DateTime Related
-function lazy_vars.CURRENT_YEAR()
-	return os.date("%Y")
-end
-function lazy_vars.CURRENT_YEAR_SHORT()
-	return os.date("%y")
-end
-function lazy_vars.CURRENT_MONTH()
-	return os.date("%m")
-end
-function lazy_vars.CURRENT_MONTH_NAME()
-	return os.date("%B")
-end
-function lazy_vars.CURRENT_MONTH_NAME_SHORT()
-	return os.date("%b")
-end
-function lazy_vars.CURRENT_DATE()
-	return os.date("%d")
-end
-function lazy_vars.CURRENT_DAY_NAME()
-	return os.date("%A")
-end
-function lazy_vars.CURRENT_DAY_NAME_SHORT()
-	return os.date("%a")
-end
-function lazy_vars.CURRENT_HOUR()
-	return os.date("%H")
-end
-function lazy_vars.CURRENT_MINUTE()
-	return os.date("%M")
-end
-function lazy_vars.CURRENT_SECOND()
-	return os.date("%S")
-end
-function lazy_vars.CURRENT_SECONDS_UNIX()
-	return tostring(os.time())
+	namespaces[name] = opts
 end
 
--- For inserting random values
+_env_namespace("", builtin_namespace)
 
-math.randomseed(os.time())
+-- The exposed api checks for the names to avoid accidental overrides
+function Environ.env_namespace(name, opts)
+	assert(
+		#name > 0 and not (name:find("_")),
+		("You can't create a namespace with name '%s' empty nor containing _"):format(
+			name
+		)
+	)
+	assert(
+		not builtin_namespace.builtin_ns[name],
+		("You can't create a namespace with name '%s' because is one one of %s"):format(
+			name,
+			builtin_ns_names
+		)
+	)
 
-function lazy_vars.RANDOM()
-	return string.format("%06d", math.random(999999))
+	_env_namespace(name, opts)
 end
 
-function lazy_vars.RANDOM_HEX()
-	return string.format("%06x", math.random(16777216)) --16^6
-end
+function Environ:__index(key)
+	local nmsp, varname = _resolve_namespace_var(key)
+	---@diagnostic disable-next-line: need-check-nil
+	local val = nmsp.vars(varname) or ""
```

Comment:
I think it's safer to return `""` it could hide a wrong variable but will avoid breaking the contract that the variable is a string or a list. One way or the other the value has to be shown as an empty string.

Reply:
I agree, it's probably safer, but I just read up on this in the lsp-snippet-spec:
"Variables
With `$name` or `${name:default}` you can insert the value of a variable. When a variable isn’t set, its default or the empty string is inserted. When a variable is unknown (that is, its name isn’t defined) the name of the variable is inserted and it is transformed into a placeholder."
(Variable with default: `${VARIABLE: this is the ${1:default}}`)
So we'll need some mechanism to determine whether a variable is not set (is an empty string, according to my short experimentation with vscode) or is just not defined.
I think both a separate function and an overload to return nil are alright (slight tendency toward having `__index` return `string|string[]|nil`, I don't think we have to worry about incorrectly named variables in lua-code)

Reply:
Ah, true, I'll take care of the parser-modifications, no need to do that here :+1:
(And yup, dynamicNode is the way forward for variables)